### PR TITLE
Add badge for NuGet download count, again

### LINKED ISSUE #678: NuGet download count no longer works
eg. http://img.shields.io/nuget/dt/RouteJs.Mvc4.svg

I understand that this was due to #655, but it would be good to have some solution to retrieve download counts. Visual Studio shows download count in the package browser when using v3 endpoints, so the data must be available somewhere.

### PR DESCRIPTION
**Pull request by Daniel15**
[NuGet download badges giveth](https://github.com/badges/shields/pull/179), and [NuGet download badges taketh away](https://github.com/badges/shields/commit/4df74acd6ce78d480fa37d11319ba13682e482f5).

Now they're back, in ~~[pog](https://www.youtube.com/watch?v=M_UR201plc8)~~ V3 form.

I switched the NuGet badges from the `SearchAutocompleteService` to the `SearchQueryService`. The SearchQueryService returns download counts in addition to version numbers.

This works for both NuGet and MyGet:

I also verified that the version number badges still work as expected:

I tried to keep the style as close to the original code as possible, although this code could do with some modernization (like using promises rather than callbacks) at some point. I'm happy to help with that 😃 

Closes #678

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -3846,6 +3860,35 @@ function mapNugetFeed(pattern, offset, getInfo) {
...
+    getNugetData(apiUrl, repo, request, function(err, nugetData) {
+      if (err != null) {
+        badgeData.text[1] = 'inaccessible';
+        sendBadge(format, badgeData);
+        return;
+      }
+      try {
+        // Official NuGet server uses "totalDownloads" whereas MyGet uses
+        // "totaldownloads" (lowercase D). Ugh.
+        var downloads = nugetData.totalDownloads || nugetData.totaldownloads;
```
Comment (paulmelnikow) [inline server.js]: What if there are zero downloads? Would `'totalDownloads' in nugetData ? nugetData.totalDownloads : nugetData.totaldownloads` be safer?

File: try.html [hunk 2]
```diff
@@ -280,6 +280,18 @@ <h3 id="downloads"> Downloads </h3>
...
+  <td><img src='/nuget/dt/Microsoft.AspNetCore.Mvc.svg' alt=''/></td>
+  <td><code>https://img.shields.io/nuget/dt/Microsoft.AspNetCore.Mvc.svg</code></td>
+  </tr>
+  <tr><th> MyGet: </th>
+  <td><img src='/myget/mongodb/dt/MongoDB.Driver.Core.svg' alt=''/></td>
+  <td><code>https://img.shields.io/myget/mongodb/dt/MongoDB.Driver.Core.svg</code></td>
+  </tr>
+  <tr><th> MyGet tenant: </th>
+  <td><img src='/dotnet.myget/dotnet-coreclr/dt/Microsoft.DotNet.CoreCLR.svg' alt=''/></td>
+  <td><code>https://img.shields.iodotnet.myget/dotnet-coreclr/dt/Microsoft.DotNet.CoreCLR.svg</code></td>
```
Comment (paulmelnikow) [inline try.html]: That link is missing a `/`.

File: server.js [hunk 3]
```diff
@@ -3768,23 +3767,38 @@ function mapNugetFeed(pattern, offset, getInfo) {
...
-            var versions = data.data;
-            if (!includePre) {
-              // Remove prerelease versions.
-              var filteredVersions = versions.filter(function(version) {
-                return !/-/.test(version);
-              });
-              if (filteredVersions.length > 0) {
-                versions = filteredVersions;
-              }
+            data = data.data && data.data[0];
```
Comment (paulmelnikow) [inline server.js]: Could you make this a more explicit check that it's an array, and with the expected length?

**Comment by paulmelnikow** [comment]
This looks great! I have a couple minor comments and I'll run some tests when you're done addressing them. Thanks for clearing up this long-standing issue.

Would you be interested in helping maintain shields? Modernization could be good… We've started using lexical scope, template strings, and arrow functions in new code. (You're welcome to use those here if you'd like.) Promises could improve error handling and reduce some of the callback boilerplate. It would be a big change, so it's something we need to plan and execute gradually. I try to avoid changing code solely for the sake of updating the style, though sometimes it's necessary. We're being conservative with `server.js` and `try.html` to avoid conflicts with the large backlog of PRs. My current focus is tests of the vendor badges (#927), which will make reviewing PRs much faster, help contributors code faster, and tell us when things break. Would you like to review #937?

File: server.js (hunk 1 above)
Comment (Daniel15) [inline server.js]: `nugetData.totalDownloads || nugetData.totaldownloads || 0` is probably fine.

In theory, everything should be using `totalDownloads` and `totaldownloads` should not exist. Certainly there'll never be a feed in which both `totalDownloads` and `totaldownloads` exist, as many JSON parsers are not case-sensitive or have an option to disable case-sensitivity. I suspect the lowercase `totaldownloads` is simply a bug in MyGet.

**Comment by Daniel15** [comment]
Thanks for the review! The issues you mentioned should be fixed now.

> Would you be interested in helping maintain shields? 

Yeah, sure! I could try to help out 😃 

> Modernization could be good

I posted some ideas in #948 if you're interested!

> . I try to avoid changing code solely for the sake of updating the style, though sometimes it's necessary. We're being conservative with server.js and try.html to avoid conflicts with the large backlog of PRs.

We could always go through the pull requests and close out all the stale ones (where the author hasn't responded). Then go through all the current ones and see what could be merged easily. I think it's fine to tell people that code is in the process of being cleaned up and they'll need to rebase, particularly if they've got a large PR :)

> My current focus is tests of the vendor badges (#927), which will make reviewing PRs much faster, help contributors code faster, and tell us when things break

This is a great focus. In my team at work, we recently had a "quality week" where we spent the whole week improving unit/integration/functional test coverage of our app, and also improving type safety (we use [Flow](https://flow.org) for typechecking all our newer JavaScript). Tests will be great as we'll feel more confident with the changes being made. 👍

**Comment by paulmelnikow** [comment]
> … we recently had a "quality week" where we spent the whole week improving unit/integration/functional test coverage of our app, and also improving type safety …

Sounds like fun!

> … we use Flow for typechecking all our newer JavaScript …

Total aside: it seems to me that TypeScript has gotten an edge recently. Why did FB choose Flow?

> We could always go through the pull requests and close out all the stale ones (where the author hasn't responded). Then go through all the current ones and see what could be merged easily. I think it's fine to tell people that code is in the process of being cleaned up and they'll need to rebase, particularly if they've got a large PR :)

This is true. However, I tend to look at unmerged contributions as an unmined asset. There are some features in there I want to include – #409, #611, #774, #786 to name a few – and if those contributors don't respond, I'll fix them up myself. With the tests, that process should be fast and pleasant. I think it's realistic. 12 days ago there were 55 open pull requests and now there are 42, so I'm hopeful we can work through it by mid-May. I've also not wanted to merge things _too_ fast, because I don't want to be too far ahead if things break on deploy, and also because I've only been onboard a month or so.

**Comment by paulmelnikow** [comment]
Just ran these badges! It's not quite clear that these numbers are download counts:

screen shot 2017-04-19 at 2 52 24 pmscreen shot 2017-04-19 at 2 52 17 pmscreen shot 2017-04-19 at 2 51 39 pm

What do you think, could we make the left side "downloads"? If showing the repository is important, we could tack "downloads" on the right side, after the metric.

This is the npm `dt` badge:

screen shot 2017-04-19 at 2 53 47 pm

**Comment by Daniel15** [comment]
Oh, good catch! I forgot to change the label on the badge to say "downloads". I can do that tonight 😃 

> Total aside: it seems to me that TypeScript has gotten an edge recently. Why did FB choose Flow?

I remember initially, Flow did a better job at inferring types from untyped code than TypeScript did. These days, both Flow and TypeScript seem good - I don't know if one has any advantages over the other. My team use Flow mainly because Flow is built by Facebook, so we can work closely with the Flow team if we encounter any issues.  The app I work on (Ads - Power Editor) was actually the first app to ever use Flow in production. We have one of the largest JS apps at Facebook so it was a good test bed 😃

**Comment by Daniel15** [comment]
Text updated!

**Comment by kzu** [comment]
I can't find the examples in the Downloads section for NuGet now, time to bring them back? ;)

**Comment by kzu** [comment]
I'm just being eager to try this, but I still can't get this to work: https://img.shields.io/nuget/dt/Moq.svg

**Comment by Daniel15** [comment]
This change hasn't been deployed yet.

**Comment by kzu** [comment]
gentle nudge 👍

**Comment by paulmelnikow** [comment]
Not forgotten. We're waiting along with you.

**Comment by paulmelnikow** [comment]
This has been deployed. 🚀

**Pull request merged**